pkg/controller/replicaschedulingpreference/controller.go

```diff
@@ -0,0 +1,740 @@
+	go func() {
+		<-stopChan
+		s.targetInformer.Stop()
+		s.podInformer.Stop()
+		s.workQueue.ShutDown()
+		s.deliverer.Stop()
+		s.clusterDeliverer.Stop()
+	}()
+}
+
+type reconciliationStatus int
+
+const (
+	statusAllOK reconciliationStatus = iota
+	statusNeedsRecheck
+	statusError
+	statusNotSynced
+)
+
+func (s *ReplicaSchedulingPreferenceController) worker() {
+	for {
+		obj, quit := s.workQueue.Get()
+		if quit {
+			return
+		}
+
+		item := obj.(*util.DelayingDelivererItem)
+		qualifiedName := item.Value.(*util.QualifiedName)
+		status := s.reconcile(*qualifiedName)
+		s.workQueue.Done(item)
+
+		switch status {
+		case statusAllOK:
+			break
+		case statusError:
+			s.deliver(*qualifiedName, 0, true)
+		case statusNeedsRecheck:
+			s.deliver(*qualifiedName, s.reviewDelay, false)
+		case statusNotSynced:
+			s.deliver(*qualifiedName, s.clusterAvailableDelay, false)
+		}
+	}
+}
+
+func (s *ReplicaSchedulingPreferenceController) deliverObj(obj pkgruntime.Object, delay time.Duration, failed bool) {
+	// TODO: right now this works on the RSP being same "ns/name" as the
+	// deploymenttemplate. Update this to rsp being able to use
+	// targetRef (kind+name) with multiple kinds.
+	// For reviewers - Plan is to use Objectmeta.OwnerReferences by setting it to
+	// created rsp into the target object. This way when reconcile is called
+	// the information will be available to link objects both ways.
+	qualifiedName := util.NewQualifiedName(obj)
+	s.deliver(qualifiedName, delay, failed)
+}
+
+// Adds backoff to delay if this delivery is related to some failure. Resets backoff if there was no failure.
+func (s *ReplicaSchedulingPreferenceController) deliver(qualifiedName util.QualifiedName, delay time.Duration, failed bool) {
+	key := qualifiedName.String()
+	if failed {
+		s.backoff.Next(key, time.Now())
+		delay = delay + s.backoff.Get(key)
+	} else {
+		s.backoff.Reset(key)
+	}
+	s.deliverer.DeliverAfter(key, &qualifiedName, delay)
+}
+
+// Check whether all data stores are in sync. False is returned if any of the informer/stores is not yet
+// synced with the corresponding api server.
+func (s *ReplicaSchedulingPreferenceController) isSynced() bool {
+	if !s.targetInformer.ClustersSynced() {
+		glog.V(2).Infof("Cluster list not synced")
+		return false
+	}
+	clusters, err := s.targetInformer.GetReadyClusters()
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Failed to get ready clusters: %v", err))
+		return false
+	}
+	if !s.targetInformer.GetTargetStore().ClustersSynced(clusters) {
+		return false
+	}
+
+	if !s.podInformer.ClustersSynced() {
+		glog.V(2).Infof("Cluster list not synced")
+		return false
+	}
+	clusters, err = s.podInformer.GetReadyClusters()
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Failed to get ready clusters: %v", err))
+		return false
+	}
+	if !s.podInformer.GetTargetStore().ClustersSynced(clusters) {
+		return false
+	}
+
+	if !s.controller.HasSynced() ||
```

Comment:
(No action required) Code golf

```golang
return [boolean expression]

// this is a potentially more readable version
controllersSynced := [boolean expression]
return controllersSynced
```

In any case, I think it's preferable to avoid using an `if` conditional to decide between `true` or `false`.  The conditional is already evaluating to a boolean, just use it directly.

Reply:
Thanks for the example; my past coding style however has always been use of `if` statements. I found them to be more readable. Meanwhile suggestion well taken, will try to adapt in future code.